### PR TITLE
persist: make state size proportional to non-empty batches

### PR DESCRIPTION
**Pull request by danhhz**
Sneaky optimization! It was previously proportional to the number of
batches (empty or non-empty). If there are no keys in a batch, we can
simply extend the description of the last one (assuming there is one).
This will be an absurdly common case in actual usage because empty
batches are used to communicate progress and most things will be
low-traffic.

Extension of work in https://github.com/MaterializeInc/materialize/pull/12225. After this final fix, a test writing 64_800
batches, mostly empty but with 1KiB of data every 1_000, into file blob
and sqlite consensus used 260KB (66 files, exactly one per batch) in
blob and 27KB in consensus (down from 7MB after 12225 and 80GB before
this work). The runtime of the test also went from ~800s after the
previous PR down to ~30s (unmeasured before this work).
### Motivation

  * This PR adds a feature that has not yet been specified.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/persist-client/src/impl/state.rs [hunk 1]
```diff
@@ -201,6 +197,26 @@ where
         }
         self.since = since;
     }
+
+    fn push_batch(&mut self, keys: &[String], desc: &Description<T>) {
```
Comment (danhhz) [inline src/persist-client/src/impl/state.rs:201]: this impl is just a placeholder. let's finish hashing out what specifically we'd like this to do

File: src/persist-client/src/impl/state.rs [hunk 2]
```diff
@@ -316,6 +332,20 @@ where
...
+// Test helpers.
+#[cfg(test)]
+impl<K, V, T, D> State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    pub fn num_trace_batches(&self) -> usize {
```
Comment (danhhz) [inline src/persist-client/src/impl/state.rs:344]: I dislike this helper. maybe I can move the test to mod state?

File: src/persist-client/src/impl/state.rs [hunk 3]
```diff
@@ -201,6 +197,26 @@ where
...
     }
+
+    fn push_batch(&mut self, keys: &[String], desc: &Description<T>) {
+        // Sneaky optimization! If there are no keys in this batch, we can
+        // simply extend the description of the last one (assuming there is
+        // one). This will be an absurdly common case in actual usage because
+        // empty batches are used to communicate progress and most things will
+        // be low-traffic.
+        if keys.is_empty() {
+            if let Some((_, last_desc)) = self.trace.last_mut() {
```
Comment (aljoscha) [inline src/persist-client/src/impl/state.rs:208]: Didn't you mention that it only extends the upper of a batch if it doesn't have keys, meaning we should check whether what we get from `last_mut()` here is actually empty?

File: src/persist-client/src/impl/state.rs [hunk 4]
```diff
@@ -105,9 +105,7 @@ where
             return Ok(Ok(()));
         }
 
-        self.trace.push((keys.to_vec(), desc.clone()));
-        debug_assert_eq!(&self.upper(), desc.upper());
-
+        self.push_batch(keys, &desc);
```
Comment (aljoscha) [inline src/persist-client/src/impl/state.rs:108]: This in itself is already a nice simplification. Small thing, but ... :raised_hands:

File: src/persist-client/src/impl/state.rs (hunk 2 above)
Comment (aljoscha) [reply src/persist-client/src/impl/state.rs:344]: I think the test can be in `mod state`, because the optimization only affects state itself.

**Review by aljoscha — APPROVED** [review]
I think this is good to merge! We'll probably keep mulling this over but I think it's a fine optimization now and helps maintaining a good image for persist. :sweat_smile:

**Review by danhhz — COMMENTED** [review]
TFTR! force pushing because I also rebased it

**Comment by aljoscha** [comment]
My approval still stands, if ci is letting you merge. :sweat_smile:

**Pull request merged**